// @1771technologies/lytenyte-core/dist/root/constants.js
export const EMPTY_POSITION_ARRAY = new Uint32Array();
export const AnyArray = Object.freeze([]);
export const AnyObject = Object.freeze({});
export const AnySet = Object.freeze(new Set());
export const DEFAULT_ROW_SOURCE = {
    useBottomCount: () => 0,
    useTopCount: () => 0,
    useRowCount: () => 0,
    useRows: () => ({ get: () => null, size: 0 }),
    useMaxRowGroupDepth: () => 0,
    useSelectionState: () => ({ kind: "isolated", selected: false, exceptions: new Set() }),
    rowsSelected: () => ({ rows: [], state: { kind: "isolated", selected: false, exceptions: new Set() } }),
    rowIndexToRowId: () => null,
    rowByIndex: () => ({ get: () => null, useValue: () => null }),
    rowById: () => null,
    rowChildren: () => [],
    rowIdToRowIndex: () => null,
    rowSiblings: () => [],
    rowInvalidate: () => { },
    rowIsSelected: () => false,
    rowLeafs: () => [],
    rowParents: () => [],
    rowsBetween: () => [],
    rowSelectionState: () => null,
    rowGroupExpansionChange: () => { },
    onRowsUpdated: () => { },
    onRowsSelected: () => { },
    onViewChange: () => { },
};
