// @1771technologies/lytenyte-core/dist/root/contexts/drop-accept.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createContext, memo, useContext, useMemo } from "react";
import { useGridIdContext } from "./grid-id.js";
const context = createContext([]);
const contextDragEvents = createContext(null);
export const DropAcceptProvider = memo((props) => {
    const gridId = useGridIdContext();
    const dropAccept = useMemo(() => {
        const drop = props.rowDropAccept ?? [];
        if (!drop.includes(gridId))
            drop.push(gridId);
        return drop.map((x) => `grid:${x}`);
    }, [gridId, props.rowDropAccept]);
    const dragValue = useMemo(() => {
        return {
            onRowDragLeave: props.onRowDragLeave,
            onRowDragEnter: props.onRowDragEnter,
            onRowDrop: props.onRowDrop,
        };
    }, [props.onRowDragEnter, props.onRowDragLeave, props.onRowDrop]);
    return (_jsx(context.Provider, { value: dropAccept, children: _jsx(contextDragEvents.Provider, { value: dragValue, children: props.children }) }));
});
export const useDropAcceptContext = () => useContext(context);
export const useDragContext = () => useContext(contextDragEvents);
