// @1771technologies/lytenyte-core/dist/root/contexts/edit-context.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createContext, useContext, useMemo, useRef, useState, } from "react";
import { useColumnsContext } from "./columns/column-context.js";
import { useEvent } from "../../hooks/use-event.js";
const context = createContext(null);
const contextSettings = createContext(null);
export function EditProvider({ api, source, ...props }) {
    const { view, columns } = useColumnsContext();
    const [activeEdit, setActiveEdit] = useState(null);
    const [editValidation, setEditValidation] = useState(true);
    const [editData, setEditData] = useState(null);
    const changeWithInit = useEvent((value, row, column) => {
        let nextData;
        const setter = (column ?? props.columnBase)?.editSetter;
        const current = editData ?? structuredClone(row.data);
        if (setter) {
            nextData = setter({ api, column, editData: current, editValue: value, row });
        }
        else {
            const field = (column.field ?? column.id);
            if (typeof field !== "number" && typeof field !== "string")
                return false;
            nextData = Array.isArray(current) ? [...current] : { ...current };
            nextData[field] = value;
        }
        return nextData;
    });
    const editDataStateRef = useRef(editData);
    editDataStateRef.current = editData;
    const changeValue = useEvent((value, column) => {
        if (!activeEdit)
            return false;
        const row = api.rowById(activeEdit.rowId);
        if (!column || !row)
            return false;
        const nextData = changeWithInit(value, row, column);
        if (nextData === false)
            return false;
        setEditData(nextData);
        editDataStateRef.current = nextData;
        const validator = props.editRowValidatorFn;
        if (validator) {
            const result = validator({ api, editData: nextData, row });
            setEditValidation(result);
            return result;
        }
        setEditValidation(true);
        return true;
    });
    const changeData = useEvent((nextData) => {
        if (!activeEdit)
            return false;
        const row = api.rowById(activeEdit.rowId);
        if (!row)
            return false;
        setEditData(nextData);
        editDataStateRef.current = nextData;
        const validator = props.editRowValidatorFn;
        if (validator) {
            const result = validator({ api, editData: nextData, row });
            setEditValidation(result);
            return result;
        }
        setEditValidation(true);
        return true;
    });
    const cancel = useEvent(() => {
        if (!activeEdit)
            return false;
        const column = view.lookup.get(activeEdit.column);
        const row = api.rowById(activeEdit.rowId);
        if (!column || !row) {
            setActiveEdit(null);
            setEditData(null);
            setEditValidation(true);
            return false;
        }
        props.onEditCancel?.({ api, row, column, editData });
        setEditValidation(true);
        setActiveEdit(null);
        setEditData(null);
    });
    const commit = useEvent(() => {
        const editDataState = editDataStateRef.current;
        if (!activeEdit)
            return false;
        const column = view.lookup.get(activeEdit.column);
        const row = api.rowById(activeEdit.rowId);
        if (!row || !column)
            return false;
        const validator = props.editRowValidatorFn;
        if (validator) {
            const valid = validator({ api, editData: editDataState, row });
            if (valid === false || typeof valid !== "boolean") {
                props.onEditFail?.({ api, column, row, editData: editDataState, validation: valid });
                return valid;
            }
        }
        for (const c of columns) {
            const mutateEdit = c?.editMutateCommit ?? props.columnBase?.editMutateCommit;
            if (!mutateEdit)
                continue;
            mutateEdit({ api, column: c, editData: editDataState, row });
        }
        let stop = false;
        props.onEditEnd?.({
            api,
            row,
            column,
            editData: editDataState,
            preventDefault: () => {
                stop = true;
            },
        });
        if (stop)
            return true;
        const updateMap = new Map();
        updateMap.set(row, editDataState);
        source.onRowsUpdated(updateMap);
        setActiveEdit(null);
        setEditValidation(true);
        setEditData(null);
        return true;
    });
    const value = useMemo(() => {
        return {
            activeEdit,
            setActiveEdit,
            editData,
            setEditData,
            editValidation,
            changeValue,
            changeWithInit,
            changeData,
            cancel,
            commit,
        };
    }, [activeEdit, cancel, changeData, changeValue, changeWithInit, commit, editData, editValidation]);
    const settings = useMemo(() => {
        return {
            editMode: props.editMode ?? "readonly",
            editClickActivator: props.editClickActivator ?? "double-click",
        };
    }, [props.editClickActivator, props.editMode]);
    return (_jsx(context.Provider, { value: value, children: _jsx(contextSettings.Provider, { value: settings, children: props.children }) }));
}
export const useEditContext = () => useContext(context);
export const useEditSettings = () => useContext(contextSettings);
