// @1771technologies/lytenyte-core/dist/root/contexts/grid-renderer-context.js
import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import { createContext, memo, useContext, useMemo } from "react";
const context = createContext(null);
const NOOP = () => null;
export const GridRendererContext = memo((props) => {
    const value = useMemo(() => {
        return {
            FullWidthRenderer: props.rowFullWidthRenderer ?? NOOP,
            DetailRenderer: props.rowDetailRenderer ?? NOOP,
            ColumnGroupRenderer: props.columnGroupRenderer ?? HeaderGroupDefault,
            Shadows: props.slotShadows,
            RowsOverlay: props.slotRowsOverlay,
            ViewportOverlay: props.slotViewportOverlay,
        };
    }, [
        props.columnGroupRenderer,
        props.rowDetailRenderer,
        props.rowFullWidthRenderer,
        props.slotRowsOverlay,
        props.slotShadows,
        props.slotViewportOverlay,
    ]);
    return _jsx(context.Provider, { value: value, children: props.children });
});
export const useGridRenderer = () => useContext(context);
function HeaderGroupDefault({ groupPath }) {
    return _jsx(_Fragment, { children: groupPath.at(-1) });
}
