// @1771technologies/lytenyte-core/dist/components/rows/row-detail-row.js
import { jsx as _jsx } from "react/jsx-runtime";
import { SCROLL_WIDTH_VARIABLE_USE, } from "@1771technologies/lytenyte-shared";
import { useEffect, useState } from "react";
import { useGridIdContext } from "../../root/contexts/grid-id.js";
import { useYCoordinates } from "../../root/contexts/coordinates.js";
import { useAPI } from "../../root/contexts/api-provider.js";
import { useRowDetailContext, useRowDetailHeightFn } from "../../root/contexts/row-detail.js";
import { useGridRenderer } from "../../root/contexts/grid-renderer-context.js";
import { useRowSourceContext } from "../../root/contexts/row-source-provider.js";
import { useDimensionContext } from "../../root/contexts/viewport/dimensions-context.js";
import { useStyleContext } from "../../root/contexts/styles-context.js";
import { sizeFromCoord } from "@1771technologies/js-utils";
export function RowDetailRow({ layout }) {
    const source = useRowSourceContext();
    const row = source.rowByIndex(layout.rowIndex).useValue();
    const { detailExpansions } = useRowDetailContext();
    if (!row || !detailExpansions.has(row.id))
        return null;
    return _jsx(RowDetailImpl, { row: row, rowIndex: layout.rowIndex });
}
function RowDetailImpl({ row, rowIndex }) {
    const id = useGridIdContext();
    const api = useAPI();
    const styles = useStyleContext();
    const dimensions = useDimensionContext();
    const yPositions = useYCoordinates();
    const { isAuto, setDetailCache } = useRowDetailContext();
    const detailHeightFn = useRowDetailHeightFn();
    const height = detailHeightFn(row.id);
    const rowHeight = sizeFromCoord(rowIndex, yPositions) - height;
    const { DetailRenderer } = useGridRenderer();
    const [detailEl, setDetailEl] = useState(null);
    useEffect(() => {
        const first = detailEl?.firstElementChild;
        if (!first || !isAuto)
            return;
        const obs = new ResizeObserver(() => {
            setDetailCache((prev) => ({
                ...prev,
                [row.id]: first.offsetHeight,
            }));
        });
        obs.observe(first);
        return () => obs.disconnect();
    }, [detailEl?.firstElementChild, isAuto, row.id, setDetailCache]);
    return (_jsx("div", { className: styles?.detail?.className, ref: setDetailEl, role: "gridcell", style: {
            pointerEvents: "none",
            position: "absolute",
            left: 0,
            width: SCROLL_WIDTH_VARIABLE_USE,
            ...styles?.detail?.style,
        }, children: _jsx("div", { tabIndex: 0, "data-ln-gridid": id, "data-ln-row-detail": true, "data-ln-rowindex": rowIndex, style: {
                position: "sticky",
                pointerEvents: "all",
                insetInlineStart: "0px",
                marginTop: rowHeight,
                width: dimensions.innerWidth,
                height: isAuto ? "auto" : height,
            }, children: _jsx(DetailRenderer, { row: row, rowIndex: rowIndex, api: api }) }) }));
}
