// @1771technologies/lytenyte-core/dist/components/rows/row-full-width.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { forwardRef, memo } from "react";
import { VIEWPORT_WIDTH_VARIABLE_USE } from "@1771technologies/lytenyte-shared";
import { RowDetailRow } from "./row-detail-row.js";
import { useRowStyle } from "./use-row-style.js";
import { useMappedEvents } from "../../hooks/use-mapped-events.js";
import { useGridIdContext } from "../../root/contexts/grid-id.js";
import { useOffsetContext } from "../../root/contexts/grid-areas/offset-context.js";
import { useYCoordinates } from "../../root/contexts/coordinates.js";
import { useAPI } from "../../root/contexts/api-provider.js";
import { useRowDetailHeightFn } from "../../root/contexts/row-detail.js";
import { useRowSourceContext } from "../../root/contexts/row-source-provider.js";
import { useGridRenderer } from "../../root/contexts/grid-renderer-context.js";
import { useStyleContext, useStyleSettings } from "../../root/contexts/styles-context.js";
import { useGridEvents } from "../../root/contexts/events-context.js";
import { useRtlContext } from "../../root/contexts/rtl-provider.js";
import { sizeFromCoord } from "@1771technologies/js-utils";
const RowFullWidthImpl = forwardRef(function RowFullWidth({ row: layout, ...props }, forwarded) {
    const id = useGridIdContext();
    const styles = useStyleContext();
    const events = useGridEvents();
    const rtl = useRtlContext();
    const { rowAlternateAttr } = useStyleSettings();
    const { FullWidthRenderer } = useGridRenderer();
    const { topOffset, headerHeight } = useOffsetContext();
    const source = useRowSourceContext();
    const yPositions = useYCoordinates();
    const row = source.rowByIndex(layout.rowIndex).useValue();
    const detailHeightFn = useRowDetailHeightFn();
    const api = useAPI();
    const handlers = useMappedEvents(events.row, { row, api, layout });
    const rowIndex = layout.rowIndex;
    const rowPin = layout.rowPin;
    const topHeight = topOffset - headerHeight;
    const detailHeight = detailHeightFn(row.id);
    const height = sizeFromCoord(rowIndex, yPositions) - detailHeight;
    return (_jsxs("div", { className: styles?.row?.className, ...handlers, ...props, role: "row", "data-ln-gridid": id, "data-ln-rowindex": rowIndex, "data-ln-alternate": rowAlternateAttr ? layout.rowIndex % 2 === 1 : undefined, "data-ln-row": true, "data-ln-last-top-pin": layout.rowLastPinTop, "data-ln-first-bottom-pin": layout.rowFirstPinBottom, "data-ln-rowtype": "full-width", 
        /** Data attributes end */
        ref: forwarded, style: useRowStyle(yPositions, rowIndex, rowPin, topHeight, detailHeight, props.style ?? styles?.row?.style), children: [_jsx("div", { role: "gridcell", tabIndex: 0, style: {
                    width: VIEWPORT_WIDTH_VARIABLE_USE,
                    height,
                    position: "sticky",
                    right: rtl ? "0px" : undefined,
                    left: rtl ? undefined : "0px",
                    pointerEvents: "all",
                }, children: row ? _jsx(FullWidthRenderer, { api: api, row: row, rowIndex: layout.rowIndex, layout: layout }) : null }), _jsx(RowDetailRow, { layout: layout })] }));
});
export const RowFullWidth = memo(RowFullWidthImpl);
