// @1771technologies/lytenyte-core/dist/components/rows/use-row-style.js
import { useMemo } from "react";
import {} from "@1771technologies/lytenyte-shared";
import { SCROLL_WIDTH_VARIABLE_USE, VIEWPORT_WIDTH_VARIABLE_USE } from "@1771technologies/lytenyte-shared";
import { sizeFromCoord } from "@1771technologies/js-utils";
export function useRowStyle(yPositions, rowIndex, rowPin, topOffset, detailHeight, propStyles) {
    const height = sizeFromCoord(rowIndex, yPositions);
    const styles = useMemo(() => {
        const isTranslated = rowPin == null;
        // FireFox does not correctly paint translations when scrolling. This results pin columns
        // appearing and disappearing as the user horizontally scrolls. However, if we directly
        // position the rows, then it appears to produce the correct calculation.
        const styles = {
            boxSizing: "border-box",
            height: isTranslated ? 0 : height,
            width: SCROLL_WIDTH_VARIABLE_USE,
            minWidth: VIEWPORT_WIDTH_VARIABLE_USE,
            pointerEvents: "none",
            whiteSpace: "nowrap",
            display: "flex",
            position: isTranslated ? "relative" : undefined,
            top: isTranslated ? yPositions[rowIndex] - topOffset : undefined,
            "--ln-row-height": `${height - detailHeight}px`,
        };
        return { ...propStyles, ...styles };
    }, [detailHeight, height, propStyles, rowIndex, rowPin, topOffset, yPositions]);
    return styles;
}
