// @1771technologies/lytenyte-core/dist/root/hooks/use-api/api-functions/use-row-drag.js
import { useDraggable } from "../../../../dnd/use-draggable.js";
import { useEvent } from "../../../../hooks/use-event.js";
export function useUseRowDrag(api, gridId) {
    return useEvent(({ rowIndex, data, placeholder, tags }) => {
        const row = api.rowByIndex(rowIndex).get();
        if (!row)
            throw new Error("The row index provided for row drag must be a valid row");
        const rowSelect = `[data-ln-gridid="${gridId}"][data-ln-row="true"][data-ln-rowindex="${rowIndex}"]`;
        // eslint-disable-next-line react-hooks/rules-of-hooks
        return useDraggable({
            data: {
                ...data,
                ...(tags ?? {}),
                [`grid:${gridId}`]: {
                    kind: "site",
                    data: {
                        row,
                        api,
                        id: gridId,
                        get rowIndex() {
                            return api.rowIdToRowIndex(row.id);
                        },
                    },
                },
                [`grid:source`]: {
                    kind: "site",
                    data: {
                        row,
                        api,
                        id: gridId,
                        get rowIndex() {
                            return api.rowIdToRowIndex(row.id);
                        },
                    },
                },
            },
            placeholder: placeholder ?? rowSelect,
        });
    });
}
