// @1771technologies/lytenyte-core/dist/root/hooks/use-api/api-functions/use-scroll-into-view.js
import { columnScrollIntoViewValue, rowScrollIntoViewValue, } from "@1771technologies/lytenyte-shared";
import { useEvent } from "../../../../hooks/use-event.js";
export function useScrollIntoView(props, vp, view, xPositions, yPositions, rowCount, rowBottomCount, rowTopCount, headerHeightTotal) {
    return useEvent((opts) => {
        if (!vp)
            return;
        let x = undefined;
        let y = undefined;
        const col = opts.column;
        if (col != null) {
            let colIndex;
            if (typeof col === "number")
                colIndex = col;
            else if (typeof col === "string")
                colIndex = view.visibleColumns.findIndex((c) => c.id === col);
            else
                colIndex = view.visibleColumns.findIndex((c) => c.id === col.id);
            if (colIndex !== -1) {
                x = columnScrollIntoViewValue({
                    centerCount: view.centerCount,
                    startCount: view.startCount,
                    endCount: view.endCount,
                    columnIndex: colIndex,
                    columnPositions: xPositions,
                    viewport: vp,
                });
            }
        }
        const row = opts.row;
        if (row != null) {
            y = rowScrollIntoViewValue({
                bottomCount: rowBottomCount,
                topCount: rowTopCount,
                rowCount: rowCount,
                headerHeight: headerHeightTotal,
                rowIndex: row,
                rowPositions: yPositions,
                viewport: vp,
            });
        }
        vp.scrollTo({
            left: x != null ? x * (props.rtl ? -1 : 1) : undefined,
            top: y,
            behavior: opts.behavior ?? "auto",
        });
    });
}
