// @1771technologies/lytenyte-core/dist/root/contexts/viewport/dimensions-context.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createContext, memo, useContext, useMemo, useState } from "react";
import { useViewportContext } from "./viewport-context.js";
import { useIsoEffect } from "../../../hooks/use-iso-effect.js";
const defaultDimensions = {
    outerHeight: 0,
    innerHeight: 0,
    outerWidth: 0,
    innerWidth: 0,
};
const context = createContext(defaultDimensions);
export const DimensionsContext = memo((props) => {
    const { viewport: vp } = useViewportContext();
    const initialDimensions = useMemo(() => {
        if (props.initialHeight || props.initialWidth) {
            return {
                innerHeight: props.initialHeight ?? 0,
                innerWidth: props.initialWidth ?? 0,
                outerHeight: props.initialHeight ?? 0,
                outerWidth: props.initialHeight ?? 0,
            };
        }
        return defaultDimensions;
    }, [props.initialHeight, props.initialWidth]);
    const [viewportDimensions, setViewportDimensions] = useState(initialDimensions);
    useIsoEffect(() => {
        if (!vp)
            return;
        let frame = null;
        const obs = new ResizeObserver(() => {
            if (frame)
                return;
            frame = requestAnimationFrame(() => {
                setViewportDimensions({
                    outerHeight: vp.offsetHeight,
                    innerHeight: vp.clientHeight,
                    outerWidth: vp.offsetWidth,
                    innerWidth: vp.clientWidth,
                });
                frame = null;
            });
        });
        obs.observe(vp);
        return () => obs.disconnect();
    }, [vp]);
    return _jsx(context.Provider, { value: viewportDimensions, children: props.children });
});
export const useDimensionContext = () => useContext(context);
