// @1771technologies/lytenyte-core/dist/__play__/perf.play.js
import { jsxs as _jsxs, jsx as _jsx } from "react/jsx-runtime";
import { useState } from "react";
import { Grid, useClientDataSource } from "../index.js";
const columns = Array.from({ length: 50 }, (_, i) => ({
    id: `${i}`,
    // rowSpan: i === 1 ? 3 : undefined,
    width: 30,
    widthMin: 30,
    field: i,
    uiHints: {
        resizable: true,
    },
}));
const data = Array.from({ length: 30_000 }, (_, r) => {
    return Array.from({ length: 50 }, (_, i) => (i === 0 ? r : Math.round(Math.random() * 100)));
});
export default function Component() {
    const [rtl, setRtl] = useState(false);
    const ds = useClientDataSource({
        data: data,
    });
    return (_jsxs("div", { children: [_jsx("div", { children: _jsxs("button", { onClick: () => setRtl((prev) => !prev), children: ["RTL: ", rtl ? "Yes" : "No"] }) }), _jsx("div", { className: "ln-grid", style: { width: "100%", height: "90vh", border: "1px solid black" }, children: _jsx(Grid, { rtl: rtl, rowSource: ds, columns: columns }) })] }));
}
