// @1771technologies/lytenyte-core/dist/components/rows/row-sections/rows-bottom.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { forwardRef, Fragment, memo, useMemo } from "react";
import { RowChildrenDefault } from "../row-children-default.js";
import { RowsSection } from "./rows-section.js";
import { CellSelectionBottom } from "../../range-selection/cell-selection-container.js";
import { useGridIdContext } from "../../../root/contexts/grid-id.js";
import { useRowCountsContext } from "../../../root/contexts/grid-areas/row-counts-context.js";
import { useOffsetContext } from "../../../root/contexts/grid-areas/offset-context.js";
import { useRowViewContext } from "../../../root/contexts/row-layout/row-layout-context.js";
export const RowsBottom = memo(forwardRef(function RowsBottom({ children = RowChildrenDefault, ...props }, forwarded) {
    const id = useGridIdContext();
    const rowView = useRowViewContext();
    const { topCount, centerCount, bottomCount } = useRowCountsContext();
    const { bottomOffset } = useOffsetContext();
    const rows = useMemo(() => {
        const rows = [];
        for (let i = 0; i < rowView.bottom.length; i++) {
            rows.push(_jsx(Fragment, { children: children(rowView.bottom[i]) }, rowView.bottom[i].id));
        }
        return rows;
    }, [children, rowView.bottom]);
    if (bottomOffset <= 0)
        return null;
    return (_jsxs(RowsSection, { ...props, ref: forwarded, rowFirst: centerCount + topCount, rowLast: centerCount + bottomCount + topCount, role: "rowgroup", "data-ln-rows-bottom": true, "data-ln-gridid": id, style: {
            ...props.style,
            height: bottomOffset,
            boxSizing: "border-box",
            position: "sticky",
            bottom: 0,
            zIndex: 3,
            minWidth: "100%",
        }, children: [_jsx(CellSelectionBottom, {}), rows] }));
}));
