// @1771technologies/lytenyte-shared/dist/layout/create-row-layout.js
export function createRowLayout({ startCutoff, centerCutoff, endCutoff, columns, topCutoff, rowCenterCutoff, bottomCutoff, rowLookback, rowByIndex, computeColSpan, computeRowSpan, isCutoff, isFullWidth, hasSpans, }) {
    const columnCount = columns.length;
    function getColumnSectionEnd(column) {
        if (column < startCutoff)
            return startCutoff;
        if (column < centerCutoff)
            return centerCutoff;
        return endCutoff;
    }
    function getRowSectionStart(row) {
        if (row >= rowCenterCutoff)
            return bottomCutoff;
        if (row >= topCutoff)
            return topCutoff;
        return 0;
    }
    function getRowSectionEnd(row) {
        if (row < topCutoff)
            return topCutoff;
        if (row < rowCenterCutoff)
            return rowCenterCutoff;
        return bottomCutoff;
    }
    function clampColSpan(column, span) {
        return Math.max(Math.min(span, getColumnSectionEnd(column) - column), 1);
    }
    function clampRowSpan(row, span) {
        return Math.max(Math.min(span, getRowSectionEnd(row) - row), 1);
    }
    const getRowPin = (index) => {
        if (index < topCutoff)
            return "top";
        if (index >= rowCenterCutoff)
            return "bottom";
        return null;
    };
    const getColPin = (index) => {
        if (index < startCutoff)
            return "start";
        if (index >= centerCutoff)
            return "end";
        return null;
    };
    function createCell(rowIndex, colIndex, colSpan, rowSpan, root = null, isDeadCol = false, isDeadRow = false) {
        return {
            kind: "cell",
            id: columns[colIndex].id,
            rowIndex,
            colIndex,
            rowPin: getRowPin(rowIndex),
            colPin: getColPin(colIndex),
            colSpan,
            rowSpan,
            isDeadCol,
            isDeadRow,
            root,
            type: columns[colIndex].type ?? "string",
            colFirstEndPin: colIndex === endCutoff,
            colLastStartPin: colIndex === startCutoff - 1,
            rowLastPinTop: rowIndex === topCutoff - 1,
            rowFirstPinBottom: rowIndex === bottomCutoff,
            rowIsFocusRow: false,
        };
    }
    const indexToLayout = new Map();
    const idToLayout = new Map();
    const occupiedFlags = new Map();
    const occupiedRoots = new Map();
    function computeRow(index) {
        const row = rowByIndex(index);
        // If the row is not defined, we return nothing, since we have nothing to compute the layout off of.
        if (!row)
            return null;
        if (isFullWidth?.(index)) {
            const layout = {
                kind: "full-width",
                id: row.id,
                rowIndex: index,
                rowPin: getRowPin(index),
                rowFirstPinBottom: index === bottomCutoff,
                rowLastPinTop: index === topCutoff - 1,
                rowIsFocusRow: false,
            };
            indexToLayout.set(index, layout);
            idToLayout.set(row.id, layout);
            return layout;
        }
        if (hasSpans) {
            const lookbackEnd = Math.max(index - rowLookback, 0);
            const sectionStart = getRowSectionStart(index);
            for (let ri = index - 1; ri >= lookbackEnd; ri--) {
                const row = rowByIndex(ri);
                if (ri < sectionStart || isFullWidth?.(ri) || !row)
                    break;
                if (indexToLayout.has(ri))
                    continue;
                const lookbackCells = new Array(columnCount);
                const riFlags = occupiedFlags.get(ri);
                const riRoots = occupiedRoots.get(ri);
                for (let ci = 0; ci < columnCount;) {
                    if (riFlags?.[ci]) {
                        const preFilledRoot = riRoots[ci];
                        lookbackCells[ci] = createCell(preFilledRoot.rowIndex, ci, 1, 1, preFilledRoot, ci !== preFilledRoot.colIndex, true);
                        ci++;
                        continue;
                    }
                    const colSpan = clampColSpan(ci, computeColSpan?.(ri, ci) ?? 1);
                    const rowSpan = clampRowSpan(ri, computeRowSpan?.(ri, ci) ?? 1);
                    const root = createCell(ri, ci, colSpan, rowSpan);
                    lookbackCells[ci] = root;
                    for (let x = ci + 1; x < ci + colSpan; x++) {
                        lookbackCells[x] = createCell(ri, x, 1, 1, root, true, false);
                    }
                    if (rowSpan > 1) {
                        for (let x = ri + 1; x < ri + rowSpan; x++) {
                            if (isCutoff(x) || isFullWidth?.(x)) {
                                root.rowSpan = x - ri;
                                break;
                            }
                            let xFlags = occupiedFlags.get(x);
                            let xRoots = occupiedRoots.get(x);
                            if (!xFlags) {
                                xFlags = new Uint8Array(columnCount);
                                xRoots = [];
                                occupiedFlags.set(x, xFlags);
                                occupiedRoots.set(x, xRoots);
                            }
                            for (let y = ci; y < ci + colSpan; y++) {
                                xFlags[y] = 1;
                                xRoots[y] = root;
                            }
                        }
                    }
                    ci += colSpan;
                }
                const lookbackRow = {
                    kind: "row",
                    id: row.id,
                    cells: lookbackCells,
                    rowIndex: ri,
                    rowPin: getRowPin(ri),
                    rowFirstPinBottom: ri === bottomCutoff,
                    rowLastPinTop: ri === topCutoff - 1,
                    rowIsFocusRow: false,
                };
                indexToLayout.set(ri, lookbackRow);
                idToLayout.set(row.id, lookbackRow);
            }
        }
        const cells = new Array(columnCount);
        const indexFlags = occupiedFlags.get(index);
        const indexRoots = occupiedRoots.get(index);
        for (let ci = 0; ci < columnCount;) {
            if (indexFlags?.[ci]) {
                const preFilledRoot = indexRoots[ci];
                cells[ci] = createCell(preFilledRoot.rowIndex, ci, 1, 1, preFilledRoot, ci !== preFilledRoot.colIndex, true);
                ci++;
                continue;
            }
            const colSpan = clampColSpan(ci, computeColSpan?.(index, ci) ?? 1);
            const rowSpan = clampRowSpan(index, computeRowSpan?.(index, ci) ?? 1);
            const root = createCell(index, ci, colSpan, rowSpan);
            cells[ci] = root;
            for (let x = ci + 1; x < ci + colSpan; x++) {
                cells[x] = createCell(index, x, 1, 1, root, true, false);
            }
            if (rowSpan > 1) {
                for (let x = index + 1; x < index + rowSpan; x++) {
                    if (isCutoff(x) || isFullWidth?.(x)) {
                        root.rowSpan = x - index;
                        break;
                    }
                    let xFlags = occupiedFlags.get(x);
                    let xRoots = occupiedRoots.get(x);
                    if (!xFlags) {
                        xFlags = new Uint8Array(columnCount);
                        xRoots = [];
                        occupiedFlags.set(x, xFlags);
                        occupiedRoots.set(x, xRoots);
                    }
                    for (let y = ci; y < ci + colSpan; y++) {
                        xFlags[y] = 1;
                        xRoots[y] = root;
                    }
                }
            }
            ci += colSpan;
        }
        const layoutRow = {
            kind: "row",
            id: row.id,
            cells,
            rowIndex: index,
            rowPin: getRowPin(index),
            rowFirstPinBottom: index === bottomCutoff,
            rowLastPinTop: index === topCutoff - 1,
            rowIsFocusRow: false,
        };
        indexToLayout.set(index, layoutRow);
        idToLayout.set(row.id, layoutRow);
        return layoutRow;
    }
    const view = {
        clearCache: () => {
            indexToLayout.clear();
            idToLayout.clear();
            occupiedFlags.clear();
            occupiedRoots.clear();
        },
        layoutById: (id) => idToLayout.get(id),
        layoutByIndex: (index) => {
            const layout = indexToLayout.get(index);
            if (layout) {
                return layout;
            }
            return computeRow(index);
        },
        rootCell: (r, c) => {
            const row = view.layoutByIndex(r);
            if (!row || row.kind === "full-width")
                return row ?? null;
            const cell = row.cells[c];
            return cell?.root ?? cell ?? null;
        },
    };
    return view;
}
