// @1771technologies/lytenyte-core/dist/root/contexts/grid-areas/grid-sections-context.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createContext, memo, useContext, useMemo } from "react";
import { useOffsetContext } from "./offset-context.js";
import { useCutoffContext } from "./cutoff-context.js";
import { useRowCountsContext } from "./row-counts-context.js";
import { useColumnsContext } from "../columns/column-context.js";
const context = createContext(null);
export const GridSectionsContextProvider = memo((props) => {
    const offsets = useOffsetContext();
    const cutoffs = useCutoffContext();
    const counts = useRowCountsContext();
    const { view } = useColumnsContext();
    const value = useMemo(() => {
        return {
            startCount: view.startCount,
            colCenterCount: view.centerCount,
            endCount: view.endCount,
            ...offsets,
            ...cutoffs,
            ...counts,
            topRowOffset: offsets.topOffset - offsets.headerHeight,
        };
    }, [counts, cutoffs, offsets, view.centerCount, view.endCount, view.startCount]);
    return _jsx(context.Provider, { value: value, children: props.children });
});
export const useGridSectionsContext = () => useContext(context);
