// @1771technologies/lytenyte-core/dist/components/viewport/viewport.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { forwardRef, memo, useMemo, useState } from "react";
import { useCombinedRefs } from "../../hooks/use-combine-refs.js";
import { useFocusTracking } from "./use-focus-tracking.js";
import { getNearestFocusable, getPositionFromFocusable, navigator, queryCell, } from "@1771technologies/lytenyte-shared";
import { beginEditing } from "./begin-editing.js";
import { RowDragMonitor } from "./row-drag-monitor.js";
import { ViewMonitor } from "./view-monitor.js";
import { useMappedEvents } from "../../hooks/use-mapped-events.js";
import { EditDriver } from "./edit-driver.js";
import { useGridIdContext } from "../../root/contexts/grid-id.js";
import { useFocusNonReactive } from "../../root/contexts/focus-position.js";
import { useOffsetContext } from "../../root/contexts/grid-areas/offset-context.js";
import { useXCoordinates, useYCoordinates } from "../../root/contexts/coordinates.js";
import { useAPI } from "../../root/contexts/api-provider.js";
import { useEditContext, useEditSettings } from "../../root/contexts/edit-context.js";
import { useRowCountsContext } from "../../root/contexts/grid-areas/row-counts-context.js";
import { useColumnsContext } from "../../root/contexts/columns/column-context.js";
import { useViewportContext } from "../../root/contexts/viewport/viewport-context.js";
import { useDimensionContext } from "../../root/contexts/viewport/dimensions-context.js";
import { useHeaderLayoutContext } from "../../root/contexts/header-layout.js";
import { useGridRenderer } from "../../root/contexts/grid-renderer-context.js";
import { useGridEvents } from "../../root/contexts/events-context.js";
import { useRtlContext } from "../../root/contexts/rtl-provider.js";
import { useStyleContext } from "../../root/contexts/styles-context.js";
import { useRowSelectionSettings } from "../../root/contexts/row-select-context.js";
import { runWithBackoff } from "@1771technologies/js-utils";
const noop = () => { };
function ViewportImpl({ children, ...props }, ref) {
    const id = useGridIdContext();
    const [vp, setVp] = useState(null);
    const api = useAPI();
    const selectActivator = useRowSelectionSettings().selectActivator;
    const events = useGridEvents();
    const rtl = useRtlContext();
    const styles = useStyleContext();
    const { editClickActivator, editMode } = useEditSettings();
    const { Shadows, ViewportOverlay } = useGridRenderer();
    const dimensions = useDimensionContext();
    const { totalHeaderHeight } = useHeaderLayoutContext();
    const { setViewport } = useViewportContext();
    const xPositions = useXCoordinates();
    const yPositions = useYCoordinates();
    const offsets = useOffsetContext();
    const edit = useEditContext();
    const focusActive = useFocusNonReactive();
    const [focused, vpFocused] = useFocusTracking(vp, focusActive, id);
    const shouldCapture = !focused && !vpFocused;
    const { rowCount } = useRowCountsContext();
    const { view } = useColumnsContext();
    const handlers = useMappedEvents(events.viewport, { viewport: vp, api });
    const handleNavigation = useMemo(() => {
        if (!vp)
            return () => { };
        return navigator({
            viewport: vp,
            gridId: id,
            scrollIntoView: api.scrollIntoView,
            getRootCell: api.cellRoot,
            isRowDetailExpanded: api.rowDetailExpanded,
            position: focusActive,
            downKey: "ArrowDown",
            upKey: "ArrowUp",
            nextKey: rtl ? "ArrowLeft" : "ArrowRight",
            prevKey: rtl ? "ArrowRight" : "ArrowLeft",
            endKey: "End",
            homeKey: "Home",
            pageDownKey: "PageDown",
            pageUpKey: "PageUp",
            columnCount: view.visibleColumns.length,
            rowCount,
        });
    }, [
        api.cellRoot,
        api.rowDetailExpanded,
        api.scrollIntoView,
        focusActive,
        id,
        rowCount,
        rtl,
        view.visibleColumns.length,
        vp,
    ]);
    const combined = useCombinedRefs(ref, setViewport, setVp);
    return (_jsxs(_Fragment, { children: [_jsx(RowDragMonitor, {}), vp && _jsx(ViewMonitor, { viewport: vp }), _jsxs("div", { className: styles?.viewport?.className, ...handlers, ...props, role: "grid", "aria-colcount": view.visibleColumns.length, "aria-rowcount": api.rowView().rowCount, "aria-multiselectable": (api.props().rowSelectionMode ?? "none") !== "none", tabIndex: 0, ref: combined, onClick: (ev) => {
                    const h = props.onClick ?? handlers.onClick;
                    h?.(ev);
                    if (ev.defaultPrevented)
                        return;
                    beginEditing(api, edit, focusActive.get(), editMode, editClickActivator, "single-click");
                    if (selectActivator === "single-click")
                        api.rowHandleSelect(ev);
                }, onDoubleClick: (ev) => {
                    const h = props.onClick ?? handlers.onClick;
                    h?.(ev);
                    if (ev.defaultPrevented)
                        return;
                    beginEditing(api, edit, focusActive.get(), editMode, editClickActivator, "double-click");
                    if (selectActivator === "double-click")
                        api.rowHandleSelect(ev);
                }, "data-ln-viewport": true, "data-ln-has-start": view.startCount > 0 ? true : undefined, "data-ln-gridid": id, onKeyDown: (e) => {
                    const h = props.onKeyDown ?? handlers.onKeyDown;
                    h?.(e);
                    if (e.defaultPrevented || e.isPropagationStopped() || !vp)
                        return;
                    const isEditing = edit.activeEdit;
                    let skipOnEdit = false;
                    const pos = focusActive.get();
                    if (isEditing) {
                        const nearestFocusable = getNearestFocusable(id, document.activeElement);
                        const active = nearestFocusable ? getPositionFromFocusable(id, nearestFocusable) : null;
                        if (pos?.kind === active?.kind &&
                            active?.kind === "cell" &&
                            pos?.kind === "cell" &&
                            pos.rowIndex === active.rowIndex &&
                            view.visibleColumns[pos.colIndex].id === isEditing.column) {
                            skipOnEdit = true;
                        }
                    }
                    // We need to cycle on tab.
                    const cycle = editMode === "row" &&
                        isEditing &&
                        pos?.kind === "cell" &&
                        api.rowByIndex(pos.rowIndex).get()?.id === isEditing.rowId;
                    if (e.key === "Tab" && isEditing && !cycle)
                        return;
                    if (cycle && e.key === "Tab") {
                        if (pos.colIndex === view.visibleColumns.length - 1 && !e.shiftKey) {
                            api.scrollIntoView({ column: 0, behavior: "instant" });
                            runWithBackoff(() => {
                                const el = queryCell(id, pos.rowIndex, 0, vp);
                                if (!el)
                                    return false;
                                el.focus();
                                return true;
                            }, [8, 16, 32, 64, 128]);
                            e.preventDefault();
                            e.stopPropagation();
                        }
                        if (pos.colIndex === 0 && e.shiftKey) {
                            api.scrollIntoView({ column: view.visibleColumns.length - 1, behavior: "instant" });
                            runWithBackoff(() => {
                                const el = queryCell(id, pos.rowIndex, view.visibleColumns.length - 1, vp);
                                if (!el)
                                    return false;
                                el.focus();
                                return true;
                            }, [8, 16, 32, 64, 128]);
                            e.preventDefault();
                            e.stopPropagation();
                        }
                        return;
                    }
                    if (!skipOnEdit) {
                        handleNavigation(e, false);
                    }
                    if (!isEditing && editMode !== "readonly") {
                        if (e.key === "Enter") {
                            beginEditing(api, edit, focusActive.get(), editMode, editClickActivator);
                            return;
                        }
                        if (e.key.length === 1 && e.key !== " ") {
                            beginEditing(api, edit, focusActive.get(), editMode, editClickActivator, undefined, e.key, true);
                            return;
                        }
                    }
                    if (isEditing) {
                        if (e.key === "Enter") {
                            const active = document.activeElement;
                            const focusable = active ? getNearestFocusable(id, active) : null;
                            if (focusable) {
                                const validation = edit.editValidation;
                                if (validation !== true)
                                    return;
                                const rowIndex = Number.parseInt(focusable.getAttribute("data-ln-rowindex"));
                                if (rowCount - 1 == rowIndex) {
                                    runWithBackoff(() => {
                                        if (focusable?.getAttribute("data-ln-edit-active") === "true")
                                            return false;
                                        focusable?.focus();
                                        return true;
                                    }, [0, 20, 40]);
                                    edit.commit();
                                }
                                else {
                                    // We don't commit the edit when moving down, since the edit will be committed when the
                                    // cell position changes.
                                    handleNavigation({
                                        key: "ArrowDown",
                                        ctrlKey: false,
                                        metaKey: false,
                                        shiftKey: false,
                                        preventDefault: noop,
                                        stopPropagation: noop,
                                    }, true);
                                }
                            }
                            return;
                        }
                        if (e.key === "Escape") {
                            const active = document.activeElement;
                            const focusable = active ? getNearestFocusable(id, active) : null;
                            edit.cancel();
                            if (focusable)
                                runWithBackoff(() => {
                                    if (focusable?.getAttribute("data-ln-edit-active") === "true")
                                        return false;
                                    focusable?.focus();
                                    return true;
                                }, [0, 20, 40]);
                            return;
                        }
                    }
                }, style: {
                    position: "relative",
                    display: "flex",
                    flexDirection: "column",
                    contain: "strict",
                    width: "100%",
                    height: "100%",
                    overflowX: "auto",
                    overflowY: "auto",
                    boxSizing: "border-box",
                    direction: rtl ? "rtl" : "ltr",
                    "--ln-vp-height": `${dimensions.innerHeight}px`,
                    "--ln-vp-row-height": `${dimensions.innerHeight - totalHeaderHeight}px`,
                    "--ln-vp-width": `${dimensions.innerWidth}px`,
                    "--ln-start-offset": `${offsets.startOffset}px`,
                    "--ln-end-offset": `${offsets.endOffset}px`,
                    "--ln-top-offset": `${offsets.topOffset}px`,
                    "--ln-bottom-offset": `${offsets.bottomOffset}px`,
                    "--ln-full-width": `${xPositions.at(-1)}px`,
                    "--ln-full-height": `${yPositions.at(-1)}px`,
                    ...(props.style ?? styles?.viewport?.style),
                }, children: [Shadows && _jsx(Shadows, {}), dimensions.innerHeight !== 0 && (_jsx(_Fragment, { children: typeof ViewportOverlay === "function" ? _jsx(ViewportOverlay, { api: api }) : ViewportOverlay })), _jsx(EditDriver, {}), vp && dimensions.innerWidth > 0 && children, dimensions.innerWidth === 0 && (_jsx("div", { suppressHydrationWarning: true, style: { minWidth: xPositions.at(-1), minHeight: yPositions.at(-1) } }))] }), shouldCapture && (_jsx("div", { role: "none", "data-ln-focus-capture": true, onFocusCapture: (e) => {
                    e.preventDefault();
                    e.stopPropagation();
                    vp?.focus();
                }, tabIndex: 0 }))] }));
}
export const Viewport = memo(forwardRef(ViewportImpl));
