// @1771technologies/lytenyte-core/dist/root/contexts/cell-range-selection/cell-range-selection-state.js
import { jsx as _jsx } from "react/jsx-runtime";
import { splitRect, } from "@1771technologies/lytenyte-shared";
import { createContext, memo, useContext, useMemo, useRef, } from "react";
import { useCutoffContext } from "../grid-areas/cutoff-context.js";
import { useControlled, useEvent, usePiece } from "../../../internal.js";
const context = createContext({});
const contextPiece = createContext(null);
const contextSettings = createContext({});
function CellSelectionContextBase(p) {
    const c = useCutoffContext();
    const [cellSelections, setCellSelections] = useControlled({
        controlled: p.cellSelections,
        default: [],
    });
    const onCellSelectionChange = useEvent((change) => {
        setCellSelections(change);
        p.onCellSelectionChange?.(change);
    });
    const cellSelections$ = usePiece(cellSelections);
    const cellSelectionsSplit = useMemo(() => {
        const splits = cellSelections.flatMap((x) => splitRect(x, c.startCutoff, c.endCutoff, c.topCutoff, c.bottomCutoff));
        return splits;
    }, [cellSelections, c.startCutoff, c.endCutoff, c.topCutoff, c.bottomCutoff]);
    const anchorRef = useRef(null);
    const settings = useMemo(() => {
        return {
            onCellSelectionChange: onCellSelectionChange,
            cellSelectionClearOnSelf: true,
            ignoreFirstColumn: p.cellSelectionExcludeMarker ?? false,
            cellSelectionMaintainOnNonCellPosition: p.cellSelectionMaintainOnNonCellPosition ?? false,
            cellSelectionMode: p.cellSelectionMode ?? "none",
            anchorRef,
        };
    }, [
        onCellSelectionChange,
        p.cellSelectionExcludeMarker,
        p.cellSelectionMaintainOnNonCellPosition,
        p.cellSelectionMode,
    ]);
    const value = useMemo(() => {
        return {
            cellSelections: cellSelections,
            cellSelectionsSplit,
        };
    }, [cellSelections, cellSelectionsSplit]);
    return (_jsx(contextSettings.Provider, { value: settings, children: _jsx(context.Provider, { value: value, children: _jsx(contextPiece.Provider, { value: cellSelections$, children: p.children }) }) }));
}
export const CellSelectionContext = memo(CellSelectionContextBase);
export const useCellRangeSelection = () => useContext(context);
export const useCellRangeSelectionSettings = () => useContext(contextSettings);
export const useCellRangeSelectionPieceContext = () => useContext(contextPiece);
