// @1771technologies/lytenyte-core/dist/data-source/hooks/use-group-tree.js
import { useMemo, useRef } from "react";
export function useGroupTree(leafs, workingSet, group, groupIdFn) {
    const groupNodeCacheRef = useRef(new Map());
    return useMemo(() => {
        const groupIdToGroupNode = new Map();
        if (!group)
            return null;
        const root = {
            kind: "root",
            children: new Map(),
            groupLookup: groupIdToGroupNode,
            maxDepth: 0,
        };
        for (let i = 0; i < workingSet.length; i++) {
            const n = leafs[workingSet[i]];
            const paths = group(n);
            let current = root.children;
            let currentGroup = root;
            // This has been marked a non-terminal node
            if (!paths?.length) {
                current.set(current.size, { kind: "leaf", row: n, parent: root, key: current.size });
                continue;
            }
            root.maxDepth = Math.max(paths.length, root.maxDepth);
            for (let j = 0; j < paths.length; j++) {
                const p = paths[j];
                if (!current.has(p)) {
                    const partial = paths.slice(0, j + 1);
                    const groupId = groupIdFn(partial);
                    const children = new Map();
                    const isLast = j === paths.length - 1;
                    if (!groupNodeCacheRef.current.get(groupId))
                        groupNodeCacheRef.current.set(groupId, {
                            __children: children,
                            last: isLast,
                            kind: "branch",
                            id: groupId,
                            data: null,
                            depth: j,
                            expandable: true,
                            expanded: false,
                            parentId: currentGroup.kind === "root" ? null : currentGroup.id,
                            key: p,
                        });
                    const node = groupNodeCacheRef.current.get(groupId);
                    node.__children = children;
                    node.last = isLast;
                    node.parentId = currentGroup.kind === "root" ? null : currentGroup.id;
                    current.set(p, {
                        id: groupId,
                        row: node,
                        kind: "branch",
                        children,
                        leafs: [],
                        leafIds: new Set(),
                        path: partial,
                        last: j === paths.length - 1,
                        parent: currentGroup,
                        key: p,
                    });
                    groupIdToGroupNode.set(groupId, current.get(p));
                }
                const node = current.get(p);
                if (node.kind !== "branch") {
                    console.error(`Invalid grouping path. Expected a group node for path: ${paths}, but found a leaf along the way.`, node);
                    continue;
                }
                node.leafs.push(i);
                node.leafIds.add(n.id);
                currentGroup = node;
                current = node.children;
            }
            const n_writable = n;
            if (currentGroup.kind === "root") {
                n_writable.depth = 0;
                n_writable.parentId = null;
            }
            else {
                n_writable.depth = currentGroup.row.depth + 1;
                n_writable.parentId = currentGroup.id;
            }
            current.set(current.size, { kind: "leaf", row: n, parent: currentGroup, key: current.size });
        }
        return root;
    }, [group, groupIdFn, leafs, workingSet]);
}
