// @1771technologies/lytenyte-shared/dist/navigation-x/attributes.js
export const getGridId = (el) => el.getAttribute("data-ln-gridid");
export const getCell = (el) => el.getAttribute("data-ln-cell");
export const getRowType = (el) => el.getAttribute("data-ln-rowtype");
export const getRow = (el) => el.getAttribute("data-ln-row");
export const getRowIndex = (el) => el.getAttribute("data-ln-rowindex");
export const getRowDetail = (el) => el.getAttribute("data-ln-row-detail");
export const getHeaderCell = (el) => el.getAttribute("data-ln-header-cell");
export const getHeaderFloating = (el) => el.getAttribute("data-ln-header-floating");
export const getHeaderRange = (el) => el.getAttribute("data-ln-header-range");
export const getHeaderGroup = (el) => el.getAttribute("data-ln-header-group");
export const getColSpan = (el) => el.getAttribute("data-ln-colspan");
export const getRowSpan = (el) => el.getAttribute("data-ln-rowspan");
export const getColIndex = (el) => el.getAttribute("data-ln-colindex");
export const getColPin = (el) => el.getAttribute("data-ln-colpin");
export const getRowPin = (el) => el.getAttribute("data-ln-rowpin");
export const getHeaderRow = (el) => el.getAttribute("data-ln-header-row");
export const getHeader = (el) => el.getAttribute("data-ln-header");
export const getViewport = (el) => el.getAttribute("data-ln-viewport");
