// @1771technologies/lytenyte-shared/dist/navigation-x/query.js
export const queryHeader = (gridId, element) => {
    return element.querySelector(`[data-ln-gridid="${gridId}"][data-ln-header="true"]`);
};
export const queryFirstFocusable = (gridId, element) => {
    const selector = `
  [data-ln-gridid="${gridId}"][data-ln-header-cell="true"][data-ln-colindex="0"],
  [data-ln-gridid="${gridId}"][data-ln-header-group="true"][data-ln-colindex="0"],
  [data-ln-gridid="${gridId}"][data-ln-header-floating="true"][data-ln-colindex="0"],
  [data-ln-gridid="${gridId}"][data-ln-cell="true"][data-ln-colindex="0"][data-ln-rowindex="0"],
  [data-ln-gridid="${gridId}"][data-ln-row="true"][data-ln-rowtype="full-width"] > div
`;
    return element.querySelector(selector);
};
export const queryCell = (gridId, rowIndex, colIndex, element) => {
    return element.querySelector(`[data-ln-gridid="${gridId}"][data-ln-cell="true"][data-ln-rowindex="${rowIndex}"][data-ln-colindex="${colIndex}"]`);
};
export const queryFullWidthRow = (gridId, rowIndex, element) => {
    return element.querySelector(`[data-ln-gridid="${gridId}"][data-ln-row="true"][data-ln-rowtype="full-width"][data-ln-rowindex="${rowIndex}"] > *`);
};
export const queryFloatingCell = (gridId, colIndex, element) => {
    return element.querySelector(`[data-ln-gridid="${gridId}"][data-ln-header-cell="true"][data-ln-header-floating=true][data-ln-colindex="${colIndex}"]`);
};
export const queryHeaderCell = (gridId, colIndex, element) => {
    return element.querySelector(`[data-ln-gridid="${gridId}"][data-ln-header-cell="true"][data-ln-colindex="${colIndex}"]`);
};
export const queryHeaderCellsAtRow = (gridId, rowIndex, element) => {
    return Array.from(element.querySelectorAll(`[data-ln-gridid="${gridId}"][data-ln-header-row-${rowIndex}="true"]`));
};
export const queryDetail = (gridId, rowIndex, element) => {
    return element.querySelector(`[data-ln-gridid="${gridId}"][data-ln-rowindex="${rowIndex}"][data-ln-row-detail="true"]`);
};
