// @1771technologies/lytenyte-core/dist/__play__/(grid-navigation)/nested-grid.play.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import "../test.css";
import { bankDataSmall } from "@1771technologies/grid-sample-data/bank-data-smaller";
import { Grid, useClientDataSource } from "../../index.js";
import "./navigation.css";
import { useState } from "react";
const baseColumns = [
    { id: "age" },
    { id: "job" },
    { id: "balance" },
    { id: "education" },
    { id: "marital" },
    { id: "default" },
    { id: "housing" },
    { id: "loan" },
    { id: "contact" },
    { id: "day" },
    { id: "month" },
    { id: "duration" },
    { id: "campaign" },
    { id: "pdays" },
    { id: "previous" },
    { id: "poutcome", name: "P Outcome" },
    { id: "y" },
];
const rowDetailRenderer = () => {
    return (_jsx("div", { style: { padding: 20, height: "100%", boxSizing: "border-box" }, children: _jsx("div", { style: { height: "100%", boxSizing: "border-box" }, children: _jsx(NestedGrid, {}) }) }));
};
export default function NestedDetailGrid({ rtl, columns, pinTop, center, pinBot, floatingRow, }) {
    const ds = useClientDataSource({
        data: center ? bankDataSmall.slice(0, center) : bankDataSmall,
        topData: pinTop ? bankDataSmall.slice(0, pinTop) : [],
        bottomData: pinBot ? bankDataSmall.slice(0, pinBot) : [],
    });
    const [expansions, setExpansions] = useState(new Set("1"));
    const [marker] = useState({
        on: true,
        cellRenderer: (p) => _jsx("button", { onClick: () => p.api.rowDetailToggle(p.row.id), children: "+" }),
    });
    return (_jsxs("div", { className: "with-border-cells", children: [_jsx("button", { tabIndex: 0, onClick: () => { }, children: "Top Capture" }), _jsx("div", { style: { width: "100%", height: "90vh", border: "1px solid black" }, children: _jsx(Grid, { gridId: "x", columns: columns ?? baseColumns, rowSource: ds, columnMarker: marker, rtl: rtl, floatingRowEnabled: floatingRow, rowDetailRenderer: rowDetailRenderer, rowDetailHeight: 400, onRowDetailExpansionsChange: setExpansions, rowDetailExpansions: expansions }) }), _jsx("button", { tabIndex: 0, onClick: () => { }, children: "Bottom Capture" })] }));
}
function NestedGrid() {
    const ds = useClientDataSource({
        data: bankDataSmall,
    });
    return _jsx(Grid, { gridId: "nested", columns: baseColumns, rowSource: ds });
}
