// @1771technologies/lytenyte-core/dist/__play__/(grid-navigation)/row-detail-with-spans.play.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import "../test.css";
import { bankDataSmall } from "@1771technologies/grid-sample-data/bank-data-smaller";
import { Grid } from "../../index.js";
import { useClientDataSource } from "../../index.js";
import "./navigation.css";
import { useState } from "react";
const baseColumns = [
    { id: "age" },
    {
        id: "job",
        rowSpan: (c) => {
            if (c.rowIndex === 0)
                return 5;
            return 1;
        },
    },
    { id: "balance" },
    { id: "education" },
    { id: "marital" },
    { id: "default" },
    { id: "housing" },
    { id: "loan" },
    { id: "contact" },
    { id: "day" },
    { id: "month" },
    { id: "duration" },
    { id: "campaign" },
    { id: "pdays" },
    { id: "previous" },
    { id: "poutcome", name: "P Outcome" },
    { id: "y" },
];
const rowDetailRenderer = () => {
    return (_jsxs("div", { children: [_jsx("button", { tabIndex: 0, children: "Detail A" }), _jsx("button", { tabIndex: 0, children: "Detail B" })] }));
};
export default function RowDetailWithSpans({ rtl, columns }) {
    const ds = useClientDataSource({
        data: bankDataSmall,
    });
    const [expansions, setExpansions] = useState(new Set(["leaf-2"]));
    const [marker] = useState({
        on: true,
        width: 60,
        cellRenderer: (p) => _jsx("button", { onClick: () => p.api.rowDetailToggle(p.row.id), children: "+" }),
    });
    return (_jsxs("div", { className: "with-border-cells", children: [_jsx("button", { tabIndex: 0, onClick: () => { }, children: "Top Capture" }), _jsx("div", { style: { width: "100%", height: "90vh", border: "1px solid black" }, children: _jsx(Grid, { gridId: "x", columns: columns ?? baseColumns, rowSource: ds, rtl: rtl, columnMarker: marker, onRowDetailExpansionsChange: setExpansions, rowDetailExpansions: expansions, rowDetailRenderer: rowDetailRenderer }) }), _jsx("button", { tabIndex: 0, onClick: () => { }, children: "Bottom Capture" })] }));
}
if (import.meta.vitest) {
    const { test, expect } = import.meta.vitest;
    const { userEvent } = await import("vitest/browser");
    const { wait, getCellQuery } = await import("../utils.js");
    const { render } = await import("vitest-browser-react");
    test("when there are row spans it should be cutoff when the detail is expanded", async () => {
        const screen = await render(_jsx(RowDetailWithSpans, {}));
        const grid = screen.getByRole("grid");
        await wait(100);
        await expect.element(grid).toBeVisible();
        await wait(); // Give the grid a moment to render
        const ourFirstCell = document.querySelector(getCellQuery("x", 0, 2));
        ourFirstCell.focus();
        await expect.element(ourFirstCell).toHaveFocus();
        await expect.element(document.activeElement).toHaveTextContent("Unemployed");
        await userEvent.keyboard("{ArrowDown}");
        await wait();
        await expect.element(document.activeElement).toHaveTextContent("Management");
        await userEvent.keyboard("{ArrowDown}");
        await wait();
        await expect.element(document.activeElement).toHaveTextContent("Detail A");
    });
}
