// @1771technologies/lytenyte-core/dist/components/header/header.js
import { jsx as _jsx } from "react/jsx-runtime";
import { forwardRef, memo, useMemo, useState } from "react";
import { useHeaderRowTemplate } from "./use-header-row-template.js";
import { useHeaderColTemplate } from "./use-header-col-template.js";
import { HeaderRowRenderer } from "./header-row/header-row-renderer.js";
import { useVirtualizedHeader } from "./use-virtualized-header.js";
import { useHeaderCellReactNodes } from "./use-header-cell-react-nodes.js";
import { HeaderProvider } from "./header-context.js";
import { useGridIdContext } from "../../root/contexts/grid-id.js";
import { useStartBoundsContext } from "../../root/contexts/bounds.js";
import { useXCoordinates } from "../../root/contexts/coordinates.js";
import { useColumnsContext } from "../../root/contexts/columns/column-context.js";
import { useHeaderHierarchyContext } from "../../root/contexts/header-hierarchy.js";
import { useHeaderLayoutContext } from "../../root/contexts/header-layout.js";
function HeaderImpl({ children = HeaderRowRenderer, ...props }, ref) {
    const id = useGridIdContext();
    const { floatingEnabled, floatingHeight, headerGroupHeight, headerHeight } = useHeaderLayoutContext();
    const columnLayout = useHeaderHierarchyContext();
    const { view } = useColumnsContext();
    const xPositions = useXCoordinates();
    const [colStartBound, colEndBound] = useStartBoundsContext();
    const gridRowTemplate = useHeaderRowTemplate(columnLayout.length, headerGroupHeight, headerHeight, floatingHeight, floatingEnabled);
    const gridColTemplate = useHeaderColTemplate(view, xPositions);
    const [active, setActive] = useState(null);
    const virtualizedHeaderCells = useVirtualizedHeader(columnLayout, active, colStartBound, colEndBound);
    const headerRows = useHeaderCellReactNodes(virtualizedHeaderCells, children);
    const value = useMemo(() => {
        return {
            activeHeaderDrag: active,
            setActiveHeaderDrag: setActive,
        };
    }, [active]);
    return (_jsx(HeaderProvider, { value: value, children: _jsx("div", { ...props, ref: ref, role: "rowgroup", "data-ln-header": true, "data-ln-gridid": id, "data-ln-rowcount": columnLayout.length - (floatingEnabled ? 1 : 0), "data-ln-floating": floatingEnabled ? true : undefined, style: {
                width: xPositions.at(-1),
                minWidth: "100%",
                boxSizing: "border-box",
                position: "sticky",
                top: 0,
                zIndex: 10,
                display: "grid",
                gridTemplateRows: gridRowTemplate,
                gridTemplateColumns: gridColTemplate,
                ...props.style,
            }, children: headerRows }) }));
}
export const Header = memo(forwardRef(HeaderImpl));
