// @1771technologies/js-utils/dist/index.js
/*
Copyright 2026 1771 Technologies

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
export { arrayShallow } from "./array-shallow.js";
export { clamp } from "./clamp.js";
export { distance } from "./distance.js";
export { equal } from "./equal.js";
export { fastShallowCompare } from "./fast-shallow-compare.js";
export { get } from "./get.js";
export { isFunction } from "./is-function.js";
export { itemsWithIdToMap } from "./items-with-id-to-map.js";
export { rangedBinarySearch } from "./ranged-binary-search.js";
export { rangesOverlap } from "./ranges-overlap.js";
export { moveRelative } from "./move-relative.js";
export { runWithBackoff } from "./run-with-backoff.js";
export { sleep, wait } from "./sleep.js";
export { makePositionArray } from "./make-position-array.js";
export { sizeFromCoord } from "./size-from-coord.js";
