// @1771technologies/lytenyte-core/dist/root/contexts/columns/column-settings-context.js
import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import { createContext, useContext, useMemo } from "react";
import { useColumnsContext } from "./column-context.js";
import { COLUMN_MARKER_ID, GROUP_COLUMN_PREFIX } from "@1771technologies/lytenyte-shared";
const context = createContext({});
export function ColumnSettingProvider({ base, children, }) {
    const columns = useColumnsContext().view.visibleColumns;
    const settings = useMemo(() => {
        const settings = columns.map((x) => {
            return [
                x.id,
                {
                    cellRenderer: x.cellRenderer ?? base?.cellRenderer ?? CellDefault,
                    editRenderer: x.editRenderer ?? base?.editRenderer ?? null,
                    editable: x.editable ?? base?.editable ?? false,
                    resizable: x.id !== COLUMN_MARKER_ID && (x.resizable ?? base?.resizable ?? false),
                    movable: (x.movable ?? base?.movable ?? false) &&
                        x.id !== COLUMN_MARKER_ID &&
                        !x.id.startsWith(GROUP_COLUMN_PREFIX),
                    headerRenderer: x.headerRenderer ?? base?.headerRenderer ?? DefaultRenderer,
                    floatingCellRenderer: x.floatingCellRenderer ?? base?.floatingCellRenderer ?? DefaultRenderer,
                    type: x.type ?? base?.type ?? "string",
                },
            ];
        });
        return Object.fromEntries(settings);
    }, [base, columns]);
    return _jsx(context.Provider, { value: settings, children: children });
}
export const useColumnSettingsContext = () => useContext(context);
function CellDefault({ column, row, api }) {
    if (row.data == null && row.loading)
        return _jsx("div", { children: "Loading..." });
    const field = api.columnField(column, row);
    return _jsx("div", { children: `${field ?? "-"}` });
}
function DefaultRenderer(p) {
    return _jsx(_Fragment, { children: p.column.name ?? p.column.id });
}
