// @1771technologies/lytenyte-core/dist/root/contexts/columns/column-context.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createContext, memo, useContext, useMemo } from "react";
import { useControlled } from "../../../hooks/use-controlled.js";
import { useEvent } from "../../../hooks/use-event.js";
import { useColumnView } from "../../hooks/use-column-view.js";
const context = createContext(null);
const EMPTY = {};
const EMPTY_COLUMNS = [];
function ColumnContextProviderBase({ children, source, ...p }) {
    const [columnGroupExpansions, setColumnGroupExpansions] = useControlled({
        controlled: p.columnGroupExpansions,
        default: EMPTY,
    });
    const onColumnGroupExpansionChange = useEvent((change) => {
        p.onColumnGroupExpansionChange?.(change);
        setColumnGroupExpansions(change);
    });
    const view = useColumnView(p, source, columnGroupExpansions);
    const value = useMemo(() => {
        return {
            view,
            columnGroupExpansions,
            columnGroupExpansionsDefault: p.columnGroupDefaultExpansion ?? true,
            columns: p.columns ?? EMPTY_COLUMNS,
            onColumnGroupExpansionChange,
            onColumnsChange: p.onColumnsChange ?? (() => { }),
        };
    }, [
        columnGroupExpansions,
        onColumnGroupExpansionChange,
        p.columnGroupDefaultExpansion,
        p.columns,
        p.onColumnsChange,
        view,
    ]);
    return _jsx(context.Provider, { value: value, children: children });
}
export const ColumnContextProvider = memo(ColumnContextProviderBase);
export const useColumnsContext = () => useContext(context);
