// @1771technologies/lytenyte-core/dist/components/range-selection/cell-selection-container.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { memo, useMemo } from "react";
import { bottomSection, centerSection, topSection } from "@1771technologies/lytenyte-shared";
import { CellSelectionRect } from "./cell-selection-rect.js";
import { useCellRangeSelection } from "../../root/contexts/cell-range-selection/cell-range-selection-state.js";
import { useCellRangeSelectionActive } from "../../root/contexts/cell-range-selection/cell-range-selection-active.js";
export const CellSelectionTop = memo(function CellSelectionTop() {
    const { cellSelectionsSplit: allRects } = useCellRangeSelection();
    const { activeSplit, deselect } = useCellRangeSelectionActive();
    const topRects = useMemo(() => {
        return allRects.filter((rect) => topSection[rect.section]);
    }, [allRects]);
    const additiveRects = useMemo(() => {
        return activeSplit.filter((rect) => topSection[rect.section]);
    }, [activeSplit]);
    return (_jsxs("div", { role: "presentation", style: {
            width: "100%",
            height: 0,
            display: "grid",
            gridTemplateRows: "0px",
            gridTemplateColumns: "0px",
        }, children: [topRects.map((rect, i) => {
                return _jsx(CellSelectionRect, { rect: rect }, i);
            }), additiveRects.map((rect) => {
                return _jsx(CellSelectionRect, { rect: rect, isDeselect: deselect }, rect.columnStart);
            })] }));
});
export const CellSelectionCenter = memo(function CellSelectionCenter() {
    const { cellSelectionsSplit: allRects } = useCellRangeSelection();
    const { activeSplit, deselect } = useCellRangeSelectionActive();
    const centerRects = useMemo(() => {
        return allRects.filter((rect) => centerSection[rect.section]);
    }, [allRects]);
    const additiveRects = useMemo(() => {
        return activeSplit.filter((rect) => centerSection[rect.section]);
    }, [activeSplit]);
    return (_jsxs("div", { role: "presentation", style: {
            width: "100%",
            height: 0,
            display: "grid",
            gridTemplateRows: "0px",
            gridTemplateColumns: "0px",
        }, children: [centerRects.map((rect, i) => {
                return _jsx(CellSelectionRect, { rect: rect }, i);
            }), additiveRects.map((rect) => {
                return _jsx(CellSelectionRect, { rect: rect, isDeselect: deselect }, rect.columnStart);
            })] }));
});
export const CellSelectionBottom = memo(function CellSelectionBottom() {
    const { cellSelectionsSplit: allRects } = useCellRangeSelection();
    const { activeSplit, deselect } = useCellRangeSelectionActive();
    const bottomRects = useMemo(() => {
        return allRects.filter((rect) => bottomSection[rect.section]);
    }, [allRects]);
    const additiveRects = useMemo(() => {
        return activeSplit.filter((rect) => bottomSection[rect.section]);
    }, [activeSplit]);
    return (_jsxs("div", { role: "presentation", style: {
            width: "100%",
            height: 0,
            display: "grid",
            gridTemplateRows: "0px",
            gridTemplateColumns: "0px",
        }, children: [bottomRects.map((rect, i) => {
                return _jsx(CellSelectionRect, { rect: rect }, i);
            }), additiveRects.map((rect) => {
                return _jsx(CellSelectionRect, { rect: rect, isDeselect: deselect }, rect.columnStart);
            })] }));
});
