// @1771technologies/lytenyte-core/dist/root/contexts/row-layout/row-layout-context.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createRowLayout, } from "@1771technologies/lytenyte-shared";
import { createContext, memo, useCallback, useContext, useMemo } from "react";
import { useCutoffContext } from "../grid-areas/cutoff-context.js";
import { useColumnsContext } from "../columns/column-context.js";
import { useViewportContext } from "../viewport/viewport-context.js";
import { useBoundsContext } from "../bounds.js";
import { useRowDetailContext } from "../row-detail.js";
import { getFullWidthFn } from "./get-full-width-fn.js";
import { getSpanFn } from "./get-span-fn.js";
const rowViewContext = createContext(null);
const rowLayoutContext = createContext(null);
export const RowLayoutProvider = memo(({ api, source, rowFullWidthPredicate, virtualizeCols, virtualizeRows, children, }) => {
    const cutoffs = useCutoffContext();
    const { view } = useColumnsContext();
    const { viewport: vp } = useViewportContext();
    const bounds = useBoundsContext();
    const { detailExpansions } = useRowDetailContext();
    const rows = source.useRows();
    const rowByIndex = useCallback((i) => {
        return rows.get(i);
    }, [rows]);
    const columns = view.visibleColumns;
    const [computeColSpan, computeRowSpan] = useMemo(() => {
        return [
            getSpanFn(rowByIndex, columns, "col", api),
            getSpanFn(rowByIndex, columns, "row", api),
        ];
    }, [api, columns, rowByIndex]);
    const isFullWidth = useMemo(() => {
        if (!rowFullWidthPredicate)
            return null;
        return getFullWidthFn(rowByIndex, rowFullWidthPredicate, api);
    }, [api, rowFullWidthPredicate, rowByIndex]);
    const isRowCutoff = useCallback((r) => {
        const row = rowByIndex(r);
        return !row || row.kind === "branch" || detailExpansions.has(row.id);
    }, [rowByIndex, detailExpansions]);
    const rowLayout = useMemo(() => {
        const l = createRowLayout({
            topCutoff: cutoffs.topCutoff,
            rowCenterCutoff: cutoffs.bottomCutoff,
            bottomCutoff: cutoffs.rowCount,
            startCutoff: cutoffs.startCutoff,
            centerCutoff: cutoffs.endCutoff,
            endCutoff: view.visibleColumns.length,
            columns: view.visibleColumns,
            computeColSpan,
            computeRowSpan,
            hasSpans: Boolean(computeColSpan || computeRowSpan),
            isCutoff: isRowCutoff,
            isFullWidth: isFullWidth,
            rowByIndex,
            rowLookback: 100,
        });
        return l;
    }, [
        cutoffs.topCutoff,
        cutoffs.bottomCutoff,
        cutoffs.rowCount,
        cutoffs.startCutoff,
        cutoffs.endCutoff,
        view.visibleColumns,
        computeColSpan,
        computeRowSpan,
        isRowCutoff,
        isFullWidth,
        rowByIndex,
    ]);
    const rowView = useMemo(() => {
        if (!vp)
            return { top: [], bottom: [], center: [], rowFirstCenter: 0, rowFocusedIndex: null };
        let n = bounds;
        if (virtualizeRows === false)
            n = { ...n, rowCenterStart: n.rowTopEnd, rowCenterEnd: n.rowCenterLast };
        if (virtualizeCols === false)
            n = { ...n, colCenterStart: n.colStartEnd, colCenterEnd: n.colCenterLast };
        const top = Array.from({ length: n.rowTopEnd }, (_, i) => rowLayout.layoutByIndex(i));
        const center = Array.from({ length: n.rowCenterEnd - n.rowCenterStart }, (_, i) => rowLayout.layoutByIndex(i + n.rowCenterStart));
        const bottom = Array.from({ length: n.rowBotEnd - n.rowBotStart }, (_, i) => rowLayout.layoutByIndex(i + n.rowBotStart));
        return {
            rowFirstCenter: n.rowCenterStart,
            rowFocusedIndex: null,
            top,
            center,
            bottom,
        };
    }, [bounds, virtualizeCols, virtualizeRows, rowLayout, vp]);
    return (_jsx(rowViewContext.Provider, { value: rowView, children: _jsx(rowLayoutContext.Provider, { value: rowLayout, children: children }) }));
});
export const useRowLayoutContext = () => useContext(rowLayoutContext);
export const useRowViewContext = () => useContext(rowViewContext);
