// @1771technologies/lytenyte-core/dist/components/rows/rows-container/rows-container.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { forwardRef, memo } from "react";
import { SCROLL_WIDTH_VARIABLE, VIEWPORT_HEIGHT_VARIABLE, VIEWPORT_WIDTH_VARIABLE, } from "@1771technologies/lytenyte-shared";
import { useRangeSelection } from "../../range-selection/use-range-selection.js";
import { useKeyboardRangeSelection } from "../../range-selection/use-keyboard-range-selection.js";
import { useCellFocusChange } from "../../range-selection/use-cell-focus-change.js";
import { useGridIdContext } from "../../../root/contexts/grid-id.js";
import { useXCoordinates, useYCoordinates } from "../../../root/contexts/coordinates.js";
import { useAPI } from "../../../root/contexts/api-provider.js";
import { useViewportContext } from "../../../root/contexts/viewport/viewport-context.js";
import { useDimensionContext } from "../../../root/contexts/viewport/dimensions-context.js";
import { useGridRenderer } from "../../../root/contexts/grid-renderer-context.js";
import { useRtlContext } from "../../../root/contexts/rtl-provider.js";
export const RowsContainer = memo(forwardRef(function Rows(props, forwarded) {
    const id = useGridIdContext();
    const rtl = useRtlContext();
    const { RowsOverlay } = useGridRenderer();
    const dimensions = useDimensionContext();
    const { viewport } = useViewportContext();
    const api = useAPI();
    const xPositions = useXCoordinates();
    const yPositions = useYCoordinates();
    const height = yPositions.at(-1);
    const width = xPositions.at(-1);
    const onMouseDown = useRangeSelection(props.onMouseDown, viewport, rtl, api);
    const onKeyDownRange = useKeyboardRangeSelection();
    useCellFocusChange();
    return (_jsxs("div", { ...props, ref: forwarded, "data-ln-rows-container": true, "data-ln-gridid": id, role: "presentation", onMouseDown: onMouseDown, onKeyDown: (e) => {
            props.onKeyDown?.(e);
            onKeyDownRange(e);
        }, style: {
            ...props.style,
            height,
            minHeight: height,
            width,
            minWidth: "100%",
            flex: "1",
            display: "flex",
            flexDirection: "column",
            userSelect: "none",
            msUserSelect: "none",
            [SCROLL_WIDTH_VARIABLE]: `${width}px`,
            [VIEWPORT_WIDTH_VARIABLE]: `${dimensions.innerWidth}px`,
            [VIEWPORT_HEIGHT_VARIABLE]: `${dimensions.innerHeight}px`,
        }, children: [dimensions.innerHeight !== 0 && (_jsx(_Fragment, { children: typeof RowsOverlay === "function" ? _jsx(RowsOverlay, { api: api }) : RowsOverlay })), props.children] }));
}));
