// @1771technologies/lytenyte-shared/dist/+constants.js
export const DEFAULT_COLUMN_WIDTH = 200;
export const DEFAULT_COLUMN_WIDTH_MAX = 1000;
export const DEFAULT_COLUMN_WIDTH_MIN = 80;
export const ROW_OVERSCAN_START = 2;
export const ROW_OVERSCAN_END = 4;
export const COL_OVERSCAN = 1;
export const FULL_WIDTH_MAP = new Map();
export const NORMAL_CELL = [1, 1];
export const DEFAULT_PREVIOUS_LAYOUT = {
    colCenterEnd: 0,
    colCenterLast: 0,
    colCenterStart: 0,
    colEndEnd: 0,
    colEndStart: 0,
    colStartEnd: 0,
    colStartStart: 0,
    rowBotEnd: 0,
    rowBotStart: 0,
    rowCenterEnd: 0,
    rowCenterLast: 0,
    rowCenterStart: 0,
    rowTopEnd: 0,
    rowTopStart: 0,
};
export const GROUP_COLUMN_PREFIX = "__ln_group__";
export const GROUP_COLUMN_MULTI_PREFIX = `${GROUP_COLUMN_PREFIX}:multi:`;
export const GROUP_COLUMN_SINGLE_ID = "__ln_group__";
export const GROUP_COLUMN_TREE_DATA = `${GROUP_COLUMN_PREFIX}:tree`;
export const COLUMN_MARKER_ID = "lytenyte-marker-column";
export const SCROLL_WIDTH_VARIABLE = "--lng-scroll-container-width";
export const SCROLL_WIDTH_VARIABLE_USE = `var(${SCROLL_WIDTH_VARIABLE})`;
export const VIEWPORT_WIDTH_VARIABLE = "--lng-viewport-width";
export const VIEWPORT_WIDTH_VARIABLE_USE = `var(${VIEWPORT_WIDTH_VARIABLE})`;
export const VIEWPORT_HEIGHT_VARIABLE = "--lng-viewport-height";
export const VIEWPORT_HEIGHT_VARIABLE_USE = `var(${VIEWPORT_HEIGHT_VARIABLE})`;
