// @1771technologies/lytenyte-core/dist/__play__/(column-header)/column-group-renderer.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
function ChevronLeft() {
    return (_jsx("svg", { xmlns: "http://www.w3.org/2000/svg", width: "12", height: "12", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2.5", strokeLinecap: "round", strokeLinejoin: "round", children: _jsx("path", { d: "M15 18l-6-6 6-6" }) }));
}
function ChevronRight() {
    return (_jsx("svg", { xmlns: "http://www.w3.org/2000/svg", width: "12", height: "12", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2.5", strokeLinecap: "round", strokeLinejoin: "round", children: _jsx("path", { d: "M9 18l6-6-6-6" }) }));
}
export function ColumnGroupRenderer({ api, collapsible, collapsed, groupPath, }) {
    return (_jsxs("div", { style: {
            display: "flex",
            alignItems: "center",
            justifyContent: "space-between",
            gap: "6px",
            width: "100%",
            height: "100%",
            padding: "0 8px",
            boxSizing: "border-box",
            fontWeight: 600,
            fontSize: "12px",
            overflow: "hidden",
        }, children: [_jsx("span", { style: { overflow: "hidden", textOverflow: "ellipsis", whiteSpace: "nowrap" }, children: groupPath.at(-1) }), collapsible && (_jsx("button", { onClick: (e) => {
                    e.stopPropagation();
                    api.columnToggleGroup(groupPath);
                }, style: {
                    flexShrink: 0,
                    display: "flex",
                    alignItems: "center",
                    justifyContent: "center",
                    width: "20px",
                    height: "20px",
                    padding: 0,
                    border: "1px solid currentColor",
                    borderRadius: "4px",
                    background: "transparent",
                    cursor: "pointer",
                    opacity: 0.7,
                }, title: collapsed ? "Expand group" : "Collapse group", children: collapsed ? _jsx(ChevronRight, {}) : _jsx(ChevronLeft, {}) }))] }));
}
