// @1771technologies/lytenyte-core/dist/hooks/use-piece.js
import { useMemo, useRef, useSyncExternalStore } from "react";
class Notifier {
    subs = new Set();
    on = (cb) => {
        this.subs.add(cb);
        return () => this.subs.delete(cb);
    };
    notify = () => {
        for (const x of this.subs)
            x();
    };
}
export function usePiece(source, setter) {
    // We use a ref to track our "Piece". A ref is nice since we can initialize it once,
    // and then never recreate the functions for the initialization. Using memos and callback
    // can achieve the same thing, however, the memo/callback will still create a function
    // even if the function is not used when the component re-renders. We completely avoid this
    // with a ref that is initialized once.
    const valueRef = useRef(null);
    if (!valueRef.current) {
        valueRef.current = {
            value: source,
            notifier: new Notifier(),
            piece: {
                get: () => valueRef.current.value,
                useValue: (selector) => {
                    const source = valueRef;
                    const snapshot = useMemo(() => {
                        if (selector)
                            return () => selector(source.current.value);
                        return () => source.current.value;
                    }, [selector, source]);
                    return useSyncExternalStore(source.current.notifier.on, snapshot, snapshot);
                },
            },
        };
    }
    // If a setter is provided we set this on the piece. Note this logic allows the setter to
    // be replaced at a later point in time. This way we can have different setters.
    if (setter)
        valueRef.current.piece.set = setter;
    if (valueRef.current.value !== source) {
        // The value has change. Hence we should notify all the subscribers.
        // However, the subscriber may have already observed the change, hence
        valueRef.current.value = source;
        queueMicrotask(valueRef.current.notifier.notify);
    }
    return valueRef.current.piece;
}
